### ISSUE TITLE
Diffusers API/Design Philosophy

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#2222]
🚨🚨 **Diffusers Design Philosophy** 🚨🚨

If you're interested in contributing or understanding this library a bit better, please have a look at the new design philosophy doc: 
https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy

Feedback and comments are very welcome below

**Comment by EandrewJones** [comment]
Great work on the document. This is much welcomed.

One area I think could use some attention is the philosophy about cross-dependencies with transformers. 

For instance, text encoders and tokenizers are important part of pipelines, especially for textual inversion and dreambooth finetuning. What if a contributor wants/needs to extend or modify one of these components to achieve some desired functionality w.r.t diffusion-specific tasks (Multivector token embeddings for textual inversion comes to mind here)? On the one hand, if you're extending functionality or creating a new class of components that come from transformers, it makes sense to keep them there (but now you have a PRs straddling two repos). On the other hand, such features may be very specific to a diffusion use-case and so making edge case changes in transformers feels off.

**TL;DR** What's the philosophy and framework you want to guide long-term thinking about diffuser's dependency on transformers?

**Comment by patrickvonplaten** [comment]
That's a very good question! In short:
- We cannot expect `transformers` to do become hacky just because diffusers needs it
- IMO it's fine to wrap some logic around transformers to make it work for special use cases, such as Multivector token embedding for textual inversion (cc @patil-suraj ). Also `transformers` can be customized quite a bit with model specific configs etc...
- It doesn't seem right to re-implement things like the tokenizer or the text model in `diffusers` just to make it more hackable for the user. The original implementation uses exactly what is used in `transformers`

**Comment by EandrewJones** [comment]
That all tracks with me. 

> IMO it's fine to wrap some logic around transformers to make it work for special use cases, such as Multivector token embedding for textual inversion (cc @patil-suraj ). Also transformers can be customized quite a bit with model specific configs etc...

 We should clarify the preferred place for this type of code in the repository.
